File: js/src/days/day03.js
import * as runner from "../lib/runner.js"
import {sum} from "../lib/iterators.js"
import { isDigit } from "../lib/parsing.js"

console.log("Solving AoC 2023 day 03")

const isSymbol = (char) => {
  return char !== "." && !isDigit(char)
}

const parseInput = (input) => {
  const lines = input.split('\n')
  const rows = []
  lines.forEach((line, rowIndex) => {
    const numbers = []
    const symbols = []
    let numberStart = null
    let numberChars = ""

    const handleDigit = (char, i) => {
      if (numberStart == null) { numberStart = i}
      numberChars += char
    }

    const handleNonDigit = (i) => {
      if (numberStart == null) { return }
      const number = {
        start: numberStart,
        end: i - 1,
        number: Number(numberChars),
        nearSymbol: false
      }

      numbers.push(number)
      numberStart = null
      numberChars = ""

      const markIfNear = (symbol) => {
        const {col} = symbol
        if (col >= number.start - 1 && col <= number.end + 1) {
          number.nearSymbol = true
        }
      }
      symbols.forEach(markIfNear)
      if (rowIndex > 0) {
        rows[rowIndex - 1].symbols.forEach(markIfNear)
      }
    }

    const handleSymbol = (char, col) => {
      const symbol = {col, char}
      symbols.push(symbol)

      const markIfNear = (number) => {
        const {start, end} = number
        if (col >= start - 1 && col <= end + 1) {
          number.nearSymbol = true
        }
      }
      numbers.forEach(markIfNear)
      if (rowIndex > 0) {
        rows[rowIndex - 1].numbers.forEach(markIfNear)
      }
    }

    for (let i = 0; i < line.length; i++) {
      const char = line[i]

      if (isDigit(char)) {
        handleDigit(char, i)
        continue
      } else {
        handleNonDigit(i)
      }
      if (isSymbol(char)) {
        handleSymbol(char, i)
      }
    } 
    handleNonDigit(line.length) // treat end of line as a non digit
    rows.push({numbers, symbols})
  })

  return rows
}

const part1 = (input) => {
  const rows = parseInput(input)
  const markedNumbers = []
  rows.forEach(({numbers}) => {
    numbers.forEach(({number, nearSymbol}) => {
      if (nearSymbol) {
        markedNumbers.push(number)
      }
    })
  })
  return sum(markedNumbers)
}

const getAdjacentNumbers = (rows, rowIndex, colIndex) => {
  const adjacent = []
  const startRow = Math.max(0, rowIndex - 1)
  const endRow = Math.min(rows.length - 1, rowIndex + 1)
  for (let i = startRow; i <= endRow; i++) {
    rows[i].numbers.forEach((number) => {
      if (colIndex >= number.start - 1 && colIndex <= number.end + 1) {
        adjacent.push(number.number)
      }
    })
  }
  return adjacent
}

const part2 = (input) => {
  const rows = parseInput(input)
  const gearRatios = []
  rows.forEach(({symbols}, rowIndex) => {
    symbols.forEach(({char, col}) => {
      if (char === "*") {
        const adjacentNumbers = getAdjacentNumbers(rows, rowIndex, col)
        if (adjacentNumbers.length === 2) {
          const [n1, n2] = adjacentNumbers
          gearRatios.push(n1 * n2)
        }
      }
    })
  })
  return sum(gearRatios)
}

await runner.testOutput('day03/example', '1', part1)
// await runner.printOutput('day03/test', part1)
// await runner.copyOutput('day03/test', part1)
// await runner.writeOutput('day03/test', '1', part1)
await runner.testOutput('day03/test', '1', part1)

await runner.testOutput('day03/example', '2', part2)
// await runner.printOutput('day03/test', part2)
// await runner.copyOutput('day03/test', part2)
// await runner.writeOutput('day03/test', '2', part2)
await runner.testOutput('day03/test', '2', part2)

File: cpp/src/days/day03.cpp
#import "day03.h"
#import "runner.h"
#import "string"
#include <sstream>

namespace {

class PartNumber {
  public:
    int value;
    int start;
    int end;
    bool nearSymbol;

    PartNumber(int value, int start, int end, bool nearSymbol)
        : value(value), start(start), end(end), nearSymbol(nearSymbol) {}
};

struct Symbol {
    char value;
    int column;
};

struct Row {
    std::vector<PartNumber> numbers;
    std::vector<Symbol> symbols;
};

bool isSymbol(char ch) { return ch != '.' && !std::isdigit(ch); }

Row parseRow(std::vector<Row> &rows, int rowIndex, const std::string &line) {
    std::vector<PartNumber> numbers;
    std::vector<Symbol> symbols;
    int numberStart = -1;
    std::string numberChars;

    auto handleDigit = [&](char ch, int column) {
        if (numberStart < 0) {
            numberStart = column;
        }
        numberChars.push_back(ch);
    };

    auto handleNonDigit = [&](int column) {
        if (numberStart >= 0) {
            int numberValue = std::stoi(numberChars);
            PartNumber newNumber(numberValue, numberStart, column - 1, false);

            // Check for nearby symbols in the current row
            for (const auto &symbol : symbols) {
                if (symbol.column >= newNumber.start - 1 && symbol.column <= newNumber.end + 1) {
                    newNumber.nearSymbol = true;
                }
            }

            // Check for nearby symbols in the previous row
            if (rowIndex > 0) {
                for (const auto &prevSymbol : rows[rowIndex - 1].symbols) {
                    if (prevSymbol.column >= newNumber.start - 1 && prevSymbol.column <= newNumber.end + 1) {
                        newNumber.nearSymbol = true;
                    }
                }
            }

            numbers.push_back(newNumber);
            numberStart = -1;
            numberChars.clear();
        }
    };

    auto handleSymbol = [&](char ch, int column) {
        Symbol newSymbol{ch, column};
        symbols.push_back(newSymbol);

        // Mark nearby numbers as having a near symbol in the current row
        for (auto &number : numbers) {
            if (column >= number.start - 1 && column <= number.end + 1) {
                number.nearSymbol = true;
            }
        }

        // Check the previous row
        if (rowIndex > 0) {
            for (auto &prevNumber : rows[rowIndex - 1].numbers) {
                if (column >= prevNumber.start - 1 && column <= prevNumber.end + 1) {
                    prevNumber.nearSymbol = true;
                }
            }
        }
    };

    for (int column = 0; column < static_cast<int>(line.length()); ++column) {
        char ch = line[column];
        if (std::isdigit(ch)) {
            handleDigit(ch, column);
        } else {
            handleNonDigit(column);
            if (isSymbol(ch)) {
                handleSymbol(ch, column);
            }
        }
    }

    // Finalize the last number in the line if any
    handleNonDigit(static_cast<int>(line.length()));
    return Row{numbers, symbols};
}

std::vector<Row> parseInput(const std::string &input) {
    std::vector<Row> rows;
    std::istringstream iss(input);
    std::string line;
    int rowIndex = 0;

    while (std::getline(iss, line)) {
        rows.push_back(parseRow(rows, rowIndex, line));
        rowIndex++;
    }

    return rows;
}

std::string part1(const std::string &input) {
    auto rows = parseInput(input);
    int result = 0;

    for (const auto &row : rows) {
        for (const auto &number : row.numbers) {
            if (number.nearSymbol) {
                result += number.value;
            }
        }
    }

    return std::to_string(result);
}

std::vector<int> getAdjacentNumberValues(const std::vector<Row> &rows, int rowIndex, int colIndex) {
    std::vector<int> adjacent;
    // Check the current row and optionally one row above and below
    int startRow = std::max(0, rowIndex - 1);
    int endRow = std::min(static_cast<int>(rows.size()) - 1, rowIndex + 1);

    for (int i = startRow; i <= endRow; ++i) {
        for (const auto &number : rows[i].numbers) {
            if (colIndex >= number.start - 1 && colIndex <= number.end + 1) {
                adjacent.push_back(number.value);
            }
        }
    }

    return adjacent;
}

std::string part2(const std::string &input) {
    auto rows = parseInput(input);
    int result = 0;

    for (size_t rowIndex = 0; rowIndex < rows.size(); ++rowIndex) {
        for (const auto &symbol : rows[rowIndex].symbols) {
            if (symbol.value == '*') {
                auto adjacentValues = getAdjacentNumberValues(rows, static_cast<int>(rowIndex), symbol.column);
                if (adjacentValues.size() == 2) {
                    result += adjacentValues[0] * adjacentValues[1];
                }
            }
        }
    }

    return std::to_string(result);
}

} // namespace

void day03() {
    testAocOutput("03", "example", "1", &part1);
    testAocOutput("03", "test", "1", &part1);
    testAocOutput("03", "example", "2", &part2);
    testAocOutput("03", "test", "2", &part2);
}
